### PR TITLE
[SYCL] Allow buitin_assume_aligned to be called from device code.

### PR DESCRIPTION
**Pull request by srividya-sundaram**
Prior to this PR, if there is a direct call to __builtin_assume_aligned from within the device code, we would get the following error:
`error: SYCL kernel cannot call a variadic function`

This PR allows `__builtin_assume_aligned` to be invoked from within device code.
The reason for doing so is, for example, if we load a float* pointer from a data structure, the compiler has no way of knowing the underlying alignment of the data the pointer points to, and therefore it cannot combine the load/stores and one ends up with many unnecessary load/store instructions plus lots of unnecessary pointer arithmetic plus increased register pressure. We see this all over the place in our kernels. With `__builtin_assume_aligned` we can tell the compiler to safely assume a certain alignment, therefore implicitly forcing coalescing.

### CODE REVIEW

File: clang/include/clang/Sema/Sema.h [hunk 1]
```diff
@@ -13975,7 +13975,7 @@ class Sema final {
     KernelConstStaticVariable
   };
 
-  bool isKnownGoodSYCLDecl(const Decl *D);
+  bool isDeclAllowedInKernel(const Decl *D);
```
Comment (Fznamznon) [inline clang/include/clang/Sema/Sema.h:13978]: Is there any justification for this renaming?
I think the old name was a bit more general.

File: clang/test/SemaSYCL/remove-restriction-builtin-assume-aligned.cpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+using namespace sycl;
+queue q;
+
+int main() {
+  int *Ptr[2];
+  // expected-no-diagnostics
+  q.submit([&](handler &h) {
+    h.single_task<class kernelA>([=]() {
+    int *APtr = (int *)__builtin_assume_aligned(Ptr, 32);
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/remove-restriction-builtin-assume-aligned.cpp:15]: Is the variadic functions handled as expected in device code? I do not remember why we prohibited this in the first place.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 3]
```diff
@@ -403,9 +403,16 @@ static bool IsSyclMathFunc(unsigned BuiltinID) {
...
 }
 
-bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
+bool Sema::isDeclAllowedInKernel(const Decl *D) {
   if (const FunctionDecl *FD = dyn_cast<FunctionDecl>(D)) {
     const IdentifierInfo *II = FD->getIdentifier();
+
+    // Allow __builtin_assume_aligned to be called from within device code.
+    if (FD->getBuiltinID() &&
+        FD->getBuiltinID() == Builtin::BI__builtin_assume_aligned) {
```
Comment (premanandrao) [inline clang/lib/Sema/SemaSYCL.cpp:412]: Nit: braces not necessary.

File: clang/test/SemaSYCL/remove-restriction-builtin-assume-aligned.cpp (hunk 2 above)
Comment (premanandrao) [reply clang/test/SemaSYCL/remove-restriction-builtin-assume-aligned.cpp:15]: Not variadic functions in general.  This builtin though is just an informational to the BE, so should be fine.   I think.

File: clang/include/clang/Sema/Sema.h (hunk 1 above)
Comment (srividya-sundaram) [reply clang/include/clang/Sema/Sema.h:13978]: `isKnownGoodSYCLDecl` didn't seem to comply with [LLVM's function naming convention](https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly)

The previous name also didn't seem to be correctly phrased. Is known what a good SYCL decl?
The new name re-purposes the same meaning in a better descriptive manner IMO.

> Function names should be verb phrases (as they represent actions), and command-like function should be imperative. The name should be camel case, and start with a lower case letter (e.g. openFile() or isFoo()).
> We cannot stress enough how important it is to use descriptive names. Pick names that match the semantics and role of the underlying entities, within reason.
>

Comment (elizabethandrews) [reply clang/include/clang/Sema/Sema.h:13978]: I read `isKnownGoodDecl` as we know this is a good/allowed decl, do not diagnose. Anyway I don't mind either name

File: clang/test/SemaSYCL/remove-restriction-builtin-assume-aligned.cpp [hunk 4]
```diff
@@ -0,0 +1,21 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -sycl-std=2020 -verify -pedantic -fsyntax-only %s
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/remove-restriction-builtin-assume-aligned.cpp:1]: Is `-pedantic `required in this test?

Comment (srividya-sundaram) [reply clang/test/SemaSYCL/remove-restriction-builtin-assume-aligned.cpp]: It isn't 'required' for this test per se, but I have used it before to make sure clang issues all warnings. Used it as an additional layer of checking if the test would emit any diagnostics, but it is not needed in hindsight.

File: clang/include/clang/Sema/Sema.h (hunk 1 above)
Comment (Fznamznon) [reply clang/include/clang/Sema/Sema.h:13978]: The new naming makes it unclear that this function in intended for SYCL. Other languages (like OpenCL) have kernels as well. Also, there is not only kernels in SYCL where we emit diagnostics.
How about something like `isDeclAllowedInSYCLDeviceCode` ?

Comment (srividya-sundaram) [reply clang/include/clang/Sema/Sema.h:13978]: I literally had the exact same name originally, but was wondering if it'd be too long a name. I'd be more than happy to go with this name. 

> Also, there is not only kernels in SYCL where we emit diagnostics.
Good point!

Comment (Fznamznon) [reply clang/include/clang/Sema/Sema.h:13978]: I think it is fine. I've seen longer names in clang.

**Comment by srividya-sundaram** [comment]
Clang format [fails](https://github.com/intel/llvm/runs/8232586851?check_suite_focus=true) on `sycl/test/basic_tests/stdcpp_compat.cpp` , a file which I haven't modified in this PR.
I did edit the files in PR directly from Github UI. Not sure if there is an issue with the Lint CI or the way I edited the files.
Any thoughts?

**Comment by srividya-sundaram** [comment]
> Clang format [fails](https://github.com/intel/llvm/runs/8232586851?check_suite_focus=true) on `sycl/test/basic_tests/stdcpp_compat.cpp` , a file which I haven't modified in this PR. I did edit the files in this PR directly from Github UI. Not sure if there is an issue with the Lint CI or the way I edited the files. Any thoughts?

Merging `sycl` branch into my branch fixed the issue.

**Review by Fznamznon — APPROVED** [review]

**Review by premanandrao — APPROVED** [review]

**Review by elizabethandrews — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]

**Comment by srividya-sundaram** [comment]
@bader This is ready for merging.

**Comment by Fznamznon** [comment]
> @bader This is ready for merging.

I think we've got several people with merge rights. I usually ping them using @intel/llvm-gatekeepers .

**Pull request merged**